File: py_ss/pendulum.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import matplotlib.pyplot as plt
import numpy as np

import blocks, helper
from blocks import N as N

class SSModel(blocks.Submodel):
    def __init__(self):
        super().__init__('pendulum')

        phi = N('phi')
        dphi = N('dphi')
        g = N('g')
        l = N('l')

        with self:
            blocks.Integrator('dphi', iport='d2phi', oport=dphi)
            blocks.Integrator('phi', iport=dphi, oport=phi, x0=np.pi/4)
            blocks.Function('sin(phi)', act_func=lambda t, x: np.sin(x),
                            iport=phi)
            blocks.MulDiv('-g/l', operations='**/',
                          iports=['-', g, l], oport='d2phi', initial=-1)

def main():
    parameters = {
        'l': 0.1,
        'g': 9.81,
        }

    model = SSModel()
    history = helper.run(model, T=np.arange(0.0, 5.0, 0.01),
                         parameters=parameters)
    
    plt.plot(history['t'], history['phi'])
    plt.show()
    
if __name__ == '__main__':
    main()

File: cpp_ss/sandbox/pendulum.cpp

#include <iostream>
#include <math.h>
#include <vector>

#include "blocks.hpp"
#include "helper.hpp"
#include "solver.hpp"
#include "gp-ios.hpp"

using namespace blocks;

class SSModel : public Submodel
{
public:
    SSModel() : Submodel("pendulum")
    {
        Node   phi(  "phi");
        Node  dphi( "dphi");
        Node d2phi("d2phi");
        Node g("g");
        Node l("l");

        enter();
        {
            new Integrator("dphi", d2phi, dphi);
            new Integrator("phi", dphi, phi, M_PI_4);
            new Function("sin(phi)",
                [](double /*t*/, const Value& x) -> Value
                {
                    return x.sin();
                }, phi);
            // new Sin("sin(phi)", {phi}, {Node()});
            new MulDiv("-g/l", "**/", {"", g, l}, d2phi, -1);
        }
        exit();
    }
};

int main()
{
    NodeValues parameters = {
        {"l", 0.1 },
        {"g", 9.81},
        };

    auto model = SSModel();
    auto history = run(model,
        [](uint k, double& t) -> bool
        {
            return arange(k, t, 0, 5, 0.01);
        },
        nullptr, parameters, rk4);

    Gnuplot gp;
	gp << "set xrange [0:500]\n";
    gp << "set yrange [-0.8:0.8]\n";
	gp << "plot" << gp.file1d(history["phi"]) << "with lines title 'x'\n";

    return 0;
}
